src/Microsoft.AspNetCore.Mvc.Analyzers/TopLevelParameterRequiredAttributeAnalyzer.cs

```diff
@@ -0,0 +1,126 @@
+﻿using System;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.AspNetCore.Mvc.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class TopLevelParameterRequiredAttributeAnalyzer : DiagnosticAnalyzer
+    {
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+            DiagnosticDescriptors.MVC1005_AttributeAvoidUsingRequiredAndBindRequired);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(compilationStartAnalysisContext =>
+            {
+                var typeCache = new SymbolCache(compilationStartAnalysisContext.Compilation);
+                if (typeCache.ControllerAttribute == null || typeCache.ControllerAttribute.TypeKind == TypeKind.Error)
```

Comment:
Here we're checking type is the Controller, in my understanding whether it is `controller` or `Model` on properties we should advice.

Reply:
> Here we're checking type is the Controller

As far as I know, this checks the `SymbolCache`, not the specific symbol. My question was why this check doesn't exist in `ModelPropertyRequiredAttribueBindRequiredAttributeAnalyzer` because that also has a `SymbolCache`.

> whether it is controller or Model on properties we should advice.

No, #7415 is specifically about top-level parameters and properties. Specifically
- model-bound properties in a controller
- parameters to an action
- model-bound properties in a page model
- parameters to a page handler

Again, ignore my optional "might be worthwhile warning" point in #7415 at least until the core work is done.

Reply:
then it should has this there as well. i will add.